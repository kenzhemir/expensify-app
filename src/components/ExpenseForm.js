import { SingleDatePicker } from "react-dates";
import moment from "moment";
import React, { Component } from "react";

import "react-dates/lib/css/_datepicker.css";

// const date = new Date();
const now = moment();
console.log(now.format("MMM Do, YYYY"));

export default class ExpenseForm extends Component {
	constructor(props) {
		super(props);
		const { description, note, amount, createdAt } = props.expense || {};
		this.state = {
			description: description || "",
			note: note || "",
			amount: amount ? (amount / 100).toString() : "",
			createdAt: createdAt ? moment(createdAt) : moment(),
			calendarFocused: false,
			error: ""
		};
	}
	onDescriptionChange = e => {
		const description = e.target.value;
		this.setState(() => ({ description }));
	};
	onAmountChange = e => {
		const amount = e.target.value;
		if (amount.match(/^\d*(\d\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	};
	onNoteChange = e => {
		const note = e.target.value;
		this.setState(() => ({ note }));
	};
	onDateChange = createdAt => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}
	};
	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }));
	};
	onSubmit = e => {
		e.preventDefault();
		if (!this.state.description || !this.state.amount) {
			this.setState({ error: "Please provide description and amount" });
		} else {
			this.setState({ error: undefined });
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			});
		}
	};
	render() {
		return (
			<div>
				{this.state.error && <p>{this.state.error}</p>}
				<form onSubmit={this.onSubmit}>
					<input
						type="text"
						autoFocus
						placeholder="Description"
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input
						type="text"
						placeholder="Amount"
						value={this.state.amount}
						onChange={this.onAmountChange}
					/>
					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
					/>
					<textarea
						name=""
						value={this.state.note}
						onChange={this.onNoteChange}
						placeholder="Add a note to your expense (optional)"
					></textarea>
					<button>Add expense</button>
				</form>
			</div>
		);
	}
}
