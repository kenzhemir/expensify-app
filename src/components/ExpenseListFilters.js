import React from "react";
import { connect } from "react-redux";
import {
	setTextFilter,
	sortByDate,
	sortByAmount,
	setStartDate,
	setEndDate
} from "../actions/filters";
import { DateRangePicker } from "react-dates";

class ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null
	};
	onDatesChange = ({ startDate, endDate }) => {
		this.props.dispatch(setStartDate(startDate));
		this.props.dispatch(setEndDate(endDate));
	};
	onFocusChange = calendarFocused => this.setState(() => ({ calendarFocused }));
	render() {
		return (
			<div>
				<input
					type="text"
					value={this.props.filters.text}
					onChange={e => this.props.dispatch(setTextFilter(e.target.value))}
				/>
				<select
					value={this.props.filters.sortBy}
					onChange={e => {
						if (e.target.value === "date") {
							this.props.dispatch(sortByDate());
						} else if (e.target.value === "amount") {
							this.props.dispatch(sortByAmount());
						}
					}}
				>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.startDate}
					startDateId="filter_date_range_start_date"
					endDate={this.props.filters.endDate}
					endDateId="filter_date_range_end_date"
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					showClearDates={true}
					isOutsideRange={() => false}
				/>
			</div>
		);
	}
}

const mapStateToProps = state => ({
	filters: state.filters
});

export default connect(mapStateToProps)(ExpenseListFilters);
