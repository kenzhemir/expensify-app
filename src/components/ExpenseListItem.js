import React from "react";

const ExpenseListItem = ({ description, amount, createdAt }) => {
	return (
		<div>
			<h3>{description}</h3>
			<p>
				{amount} - {createdAt}
			</p>
		</div>
	);
};

export default ExpenseListItem;
