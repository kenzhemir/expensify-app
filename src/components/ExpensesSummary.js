import { connect } from "react-redux";
import numeral from "numeral";
import React from "react";

import getTotalExpenses from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
	const expenseWord = expensesCount == 1 ? "expense" : "expenses";
	const formattedTotal = numeral(expensesTotal / 100).format("$0,0.00");
	return (
		<p>
			Viewing {expensesCount} {expenseWord} totalling {formattedTotal}
		</p>
	);
};

const mapStateToProps = state => {
	const visibleExpenses = selectExpenses(state.expenses, state.filters);
	return {
		expensesCount: visibleExpenses.length,
		expensesTotal: getTotalExpenses(visibleExpenses)
	};
};

export default connect(mapStateToProps)(ExpensesSummary);
