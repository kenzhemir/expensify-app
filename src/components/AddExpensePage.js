import { connect } from "react-redux";
import React from "react";

import { startAddExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";
import SimplePageHeader from "./SimplePageHeader";

export class AddExpensePage extends React.Component {
	onSubmit = expense => {
		this.props.startAddExpense(expense);
		this.props.history.push("/");
	};
	render() {
		return (
			<div>
				<SimplePageHeader title="Add Expense" />
				<div className="content-container">
					<ExpenseForm onSubmit={this.onSubmit} />
				</div>
			</div>
		);
	}
}

const mapDispatchToProps = dispatch => ({
	startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
