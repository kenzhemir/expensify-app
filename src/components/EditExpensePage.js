import { connect } from "react-redux";
import React from "react";

import { startEditExpense } from "../actions/expenses";
import { startRemoveExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";
import SimplePageHeader from "./SimplePageHeader";

export class EditExpensePage extends React.Component {
	onSubmit = expense => {
		this.props.startEditExpense(this.props.expense.id, expense);
		this.props.history.push("/");
	};
	onRemove = e => {
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push("/");
	};
	render() {
		return (
			<div>
				<SimplePageHeader title="Edit Expense" />
				<div className="content-container">
					<ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
					<button onClick={this.onRemove} className="button button--secondary">
						Remove expense
					</button>
				</div>
			</div>
		);
	}
}

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: data => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
