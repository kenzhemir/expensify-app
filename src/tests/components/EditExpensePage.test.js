import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let startEditExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
	startEditExpense = jest.fn();
	startRemoveExpense = jest.fn();
	history = { push: jest.fn() };
	wrapper = shallow(
		<EditExpensePage
			startEditExpense={startEditExpense}
			startRemoveExpense={startRemoveExpense}
			history={history}
			expense={expenses[0]}
		/>
	);
});

test("should render edit expense page correctly", () => {
	expect(wrapper).toMatchSnapshot();
});

test("should handle startEditExpense", () => {
	const updates = { ...expenses[1], id: undefined };
	wrapper.find("ExpenseForm").prop("onSubmit")(updates);
	expect(startEditExpense).toHaveBeenLastCalledWith(expenses[0].id, updates);
	expect(history.push).toHaveBeenLastCalledWith("/");
});

test("should handle startRemoveExpense", () => {
	wrapper.find("button").simulate("click");
	expect(startRemoveExpense).toHaveBeenLastCalledWith({ id: expenses[0].id });
	expect(history.push).toHaveBeenLastCalledWith("/");
});
