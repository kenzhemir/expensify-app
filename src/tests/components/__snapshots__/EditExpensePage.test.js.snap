// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page correctly 1`] = `
<div>
  <SimplePageHeader
    title="Edit Expense"
  />
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "Gum",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove expense
    </button>
  </div>
</div>
`;
